core/drivers/crypto/hisilicon/sec/hisi_sec.c

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright 2022 HiSilicon Limited.
+ * Kunpeng hardware accelerator sec module init.
+ */
+#include "hisi_sec.h"
+
+#define AM_CFG_SINGLE_PORT_MAX_TRANS 0x300014
+#define SEC_CORE_INT_MASK 0x301000
+#define SEC_CORE_INT_SOURCE 0x301010
+#define SEC_RAS_CE_ENABLE 0x301050
+#define SEC_RAS_FE_ENABLE 0x301054
+#define SEC_RAS_NFE_ENABLE 0x301058
+#define SEC_MEM_START_INIT 0x301100
+#define SEC_MEM_INIT_DONE 0x301104
+#define SEC_CONTROL_REG 0x301200
+#define SEC_INTERFACE_USER_CTRL0 0x301220
+#define SEC_INTERFACE_USER_CTRL1 0x301224
+#define SEC_SAA_EN 0x301270
+#define SEC_BD_ERR_CHK_EN0 0x301380
+#define SEC_BD_ERR_CHK_EN1 0x301384
+#define SEC_BD_ERR_CHK_EN3 0x30138c
+#define SEC_DYNAMIC_GATE_V3 0x30121c
+#define SEC_CORE_AUTO_GATE_V3 0x30212c
+#define SEC_INTERFACE_USER_CTRL0_V3 0x302220
+#define SEC_INTERFACE_USER_CTRL1_V3 0x302224
+#define SEC_SINGLE_PORT_MAX_TRANS 0x2060
+#define SEC_ABNML_INT_DISABLE 0x0
+#define SEC_RAS_CE_ENB_MASK 0x88
+#define SEC_RAS_FE_ENB_MASK 0x0
+#define SEC_RAS_NFE_ENB_MASK 0x177
+#define SEC_CLK_GATE_ENABLE BIT(3)
+#define SEC_DYNAMIC_GATE_EN 0x7bff
+#define SEC_CORE_AUTO_GATE_EN GENMASK_32(3, 0)
+#define SEC_TRNG_EN_MASK BIT(8)
+#define SEC_SAA_ENABLE 0x17f
+#define SEC_SAA_ENABLE_V3 0xf
+#define SEC_BD_ERR_CHK0 0xefffffff
+#define SEC_BD_ERR_CHK1 0x7ffff7fd
+#define SEC_BD_ERR_CHK3 0xffffbfff
+#define SEC_USER0_CFG 0x20200
+#define SEC_USER0_SMMU_NORMAL (BIT(23) | BIT(15))
+#define SEC_USER1_CFG 0x12141214
+#define SEC_USER1_SMMU_NORMAL (BIT(31) | BIT(23) | BIT(15) | BIT(7))
+#define SEC_USER0_CFG_V3 0x20200
+#define SEC_USER1_CFG_V3 0x8c494
+#define SEC_LITTLE_ENDIAN 0
+
+static SLIST_HEAD(, acc_device) sec_list = SLIST_HEAD_INITIALIZER(sec_list);
+
+struct hisi_qp *hisi_sec_create_qp(uint8_t sq_type)
+{
+	struct acc_device *sec_dev = NULL, *cur_dev = NULL;
+	uint32_t max_qp_num = 0;
+	uint32_t free_qp_num;
+	struct hisi_qm *qm;
+
+	SLIST_FOREACH(cur_dev, &sec_list, link) {
+		qm = &cur_dev->qm;
+		free_qp_num = (qm->fun_type == HISI_QM_HW_PF ?
+			HISI_QM_PF_Q_NUM : HISI_QM_VF_Q_NUM) - qm->qp_in_used;
+		if (free_qp_num > max_qp_num) {
+			max_qp_num = free_qp_num;
+			sec_dev = cur_dev;
+		}
+	}
+
+	if (!sec_dev) {
+		EMSG("No available sec device");
+		return NULL;
+	}
+
+	return hisi_qm_create_qp(&sec_dev->qm, sq_type);
+}
+
+static void sec_disable_clock_gate(struct hisi_qm *qm)
+{
+	/* HISI_QM_HW_V2 version need to close clock gating */
+	io_clrbits32(qm->io_base + SEC_CONTROL_REG, SEC_CLK_GATE_ENABLE);
+}
+
+static void sec_enable_clock_gate(struct hisi_qm *qm)
+{
+	if (qm->version == HISI_QM_HW_V2)
+		return;
+
+	io_setbits32(qm->io_base + SEC_CONTROL_REG, SEC_CLK_GATE_ENABLE);
+	io_write32(qm->io_base + SEC_DYNAMIC_GATE_V3, SEC_DYNAMIC_GATE_EN);
+	io_write32(qm->io_base + SEC_CORE_AUTO_GATE_V3, SEC_CORE_AUTO_GATE_EN);
+}
+
+static uint32_t sec_engine_init(struct acc_device *sec_dev)
+{
+	struct hisi_qm *qm = &sec_dev->qm;
+	uint32_t val = 0;
+	uint32_t ret = 0;
```

Comment:
Recommendation: `enum hisi_drv_status ret = HISI_QM_DRVCRYPT_NO_ERR`ere and at many other places.